### PR TITLE
[RyuJIT/ARM32] Unassign double register properly at BB entry

### LINKED ISSUE #12078: [RyuJIT/ARM32] Assertion failed 'assignedInterval == nextRegRec->assignedInterval'
With COMPlus_AltJit=* COMPlus_AltJitAssertOnNYI=0 options in Checked build.

One tests(`JIT/Regression/CLR-x86-JIT/V1.2-M01/b16570/b16570`) faced following assertion.
```
FAILED   - [5342][  13s]JIT/Regression/CLR-x86-JIT/V1.2-M01/b16570/b16570/b16570.sh
               BEGIN EXECUTION
               /home/dotnet_jenkins/workspace/hqueue/Run_coreclr_tests_Checked_ryujit_nyi_all/runtime/tests/Tests/coreoverlay/corerun b16570.exe

               Assert failure(PID 25238 [0x00006296], Thread: 25238 [0x6296]): Assertion failed 'assignedInterval == nextRegRec->assignedInterval' in 'test:Func():double' (IL size 395)

                   File: /opt/code/github/hqueue/coreclr/src/jit/lsra.cpp Line: 6340
                   Image: /home/dotnet_jenkins/workspace/hqueue/Run_coreclr_tests_Checked_ryujit_nyi_all/runtime/tests/Tests/coreoverlay/corerun

               ./b16570.sh: line 240: 25238 Aborted                 (core dumped) $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
               Expected: 100
               Actual: 134
               END EXECUTION - FAILED
```

### PR DESCRIPTION
**Pull request by hqueue**
Fix #12078 

When updating var locations at basic block entry,
we should un-assign double register properly for TYP_DOUBLE interval.

# Example

## Before
With `COMPlus_AltJit=* COMPlus_JitDump=test:Func` option, `JIT/Regression/CLR-x86-JIT/V1.2-M01/b16570/b16570` shows following allocation progress and raise assertion later.

```
Loc RP#  Name  Type  Action Reg  |f8   |f9   |f10  |f11  |
---------------------------------+-----+-----+-----+-----+
                                 |     |     |     |     |
  0.#0   BB1 PredBB0             |     |     |     |     |
  8.#1   C14   Def    Alloc r0   |     |     |     |     |
188.#253 V1    Def    Alloc f9   |V10 a|V1  a|C73 i|     |
...(omitted)...
209.#265 V10   Use *  Keep  f8   |V10 a|V1  a|V9  a|     |
---------------------------------+-----+-----+-----+-----+
Loc RP#  Name  Type  Action Reg  |f8   |f9   |f10  |f11  |
---------------------------------+-----+-----+-----+-----+
213.#266 BB2 PredBB1             |     |V1  a|V9  a|     |
...(omitted)...
230.#279 V8    Def    Alloc f8   |V8  a|V8  a|V9  i|     |// Interval V8 is TYP_DOUBLE.  f8 and f9 are allocated as double register (d4).
...(omitted)...
289.#337 V3    Use *  Keep  f0   |V8  i|V8  i|V9  i|     |
---------------------------------+-----+-----+-----+-----+
Loc RP#  Name  Type  Action Reg  |f8   |f9   |f10  |f11  |
---------------------------------+-----+-----+-----+-----+
291.#338 BB3 PredBB1             |V8  i|V1  a|V9  a|     |// Interval V1 is live-in at `f9` from BB1.
299.#339 V0    Use *  Keep  r4   |V8  i|V1  a|V9  a|     |
300.#340 I107  Def    Alloc r3   |V8  i|V1  a|V9  a|     |
```

BB3 has one pred BB1. BB2 returns at end, i.e.  no successor BB.

When interval `V1` is live-in at `f9` from BB1, we should unassign `f8` and `f9` together properly, because previously assigned interval `V8` is `TYP_DOUBLE`.

## After

`f8` and `f9` are un-assigned properly and interval `V1` reside in `f9` as expected.
And test is passed without problem.
```
---------------------------------+-----+-----+-----+-----+
Loc RP#  Name  Type  Action Reg  |f8   |f9   |f10  |f11  |
---------------------------------+-----+-----+-----+-----+
291.#338 BB3 PredBB1             |     |V1  a|V9  a|     |
299.#339 V0    Use *  Keep  r4   |     |V1  a|V9  a|     |
300.#340 I107  Def    Alloc r3   |     |V1  a|V9  a|     |
```

### CODE REVIEW

**Comment by BruceForstall** [comment]
@dotnet/arm32-contrib

File: src/jit/lsra.cpp [hunk 1]
```diff
@@ -7005,23 +7005,34 @@ void LinearScan::processBlockStartLocations(BasicBlock* currentBlock, bool alloc
...
-                            inVarToRegMap[assignedInterval->getVarIndex(compiler)] == targetReg)
+                            inVarToRegMap[assignedInterval->getVarIndex(compiler)] == assignedRegNum)
                         {
                             inVarToRegMap[assignedInterval->getVarIndex(compiler)] = REG_STK;
                         }
                     }
                     else
                     {
-                        // This interval is no longer assigned to this register.
+// This interval is no longer assigned to this register.
```
Comment (hqueue) [inline src/jit/lsra.cpp]: Although it look weird, jit-format forced me to format the comment like this.

Comment (CarolEidt) [inline src/jit/lsra.cpp]: You can use a `CLANG_FORMAT_COMMENT_ANCHOR;` after your comment (and aligned with it) to get around this problem.

File: src/jit/lsra.cpp [hunk 2]
```diff
@@ -7005,23 +7005,34 @@ void LinearScan::processBlockStartLocations(BasicBlock* currentBlock, bool alloc
...
+#ifdef _TARGET_ARM_
+                        updateAssignedInterval(targetRegRecord, nullptr, assignedInterval->registerType);
```
Comment (CarolEidt) [inline src/jit/lsra.cpp]: You should change `updateAssignedInterval` to be unconditionally defined, and to do the simple assignment when not `_TARGET_ARM_`. Then you eliminate a lot of `#ifdef`s

File: src/jit/lsra.cpp [hunk 3]
```diff
@@ -7005,23 +7005,34 @@ void LinearScan::processBlockStartLocations(BasicBlock* currentBlock, bool alloc
                 // Is there another interval currently assigned to this register?  If so unassign it.
                 if (assignedInterval != nullptr)
                 {
+#ifdef _TARGET_ARM_
+                    if (assignedInterval->assignedReg == targetRegRecord ||
+                        isSecondHalfReg(targetRegRecord, assignedInterval))
+#else
                     if (assignedInterval->assignedReg == targetRegRecord)
+#endif
```
Comment (CarolEidt) [inline src/jit/lsra.cpp]: Here you could define a method such as `isAssignedToInterval(Interval, RegRecord)` that could do the simple equivalence check when not `_TARGET_ARM_`.

Comment (hqueue) [inline src/jit/lsra.cpp]: I will update it as suggested. Thanks!

File: src/jit/lsra.cpp (hunk 2 above)
Comment (hqueue) [inline src/jit/lsra.cpp]: Right, I think we can remove a lot of `#ifdef` throughout `lsra.cpp`.
I will apply suggested approach and let see how it looks.

File: src/jit/lsra.cpp (hunk 1 above)
Comment (hqueue) [inline src/jit/lsra.cpp]: Thanks :)

**Comment by hqueue** [comment]
@CarolEidt PTAL. Thanks!

**Pull request merged**